src/input/singleton.rs

```diff
@@ -1,52 +1,56 @@
-use crossbeam::atomic::AtomicCell;
-use parking_lot::Mutex;
+use std::sync::atomic::{AtomicU32, Ordering};
 
-use crate::{id::FromId, Id};
+use crate::Id;
 
 mod sealed {
     pub trait Sealed {}
 }
 
 pub trait SingletonChoice: sealed::Sealed + Default {
-    fn with_lock(&self, cb: impl FnOnce() -> Id) -> Id;
+    fn with_scope(&self, cb: impl FnOnce() -> Id) -> Id;
     fn index(&self) -> Option<Id>;
 }
 
 pub struct Singleton {
-    index: AtomicCell<Option<Id>>,
-    lock: Mutex<()>,
+    index: AtomicU32,
 }
 impl sealed::Sealed for Singleton {}
 impl SingletonChoice for Singleton {
-    fn with_lock(&self, cb: impl FnOnce() -> Id) -> Id {
-        let _guard = self.lock.lock();
-        if self.index.load().is_some() {
+    fn with_scope(&self, cb: impl FnOnce() -> Id) -> Id {
+        if self.index.load(Ordering::Acquire) != 0 {
             panic!("singleton struct may not be duplicated");
         }
         let id = cb();
-        self.index.store(Some(id));
-        drop(_guard);
+        if self
+            .index
+            .compare_exchange(0, id.as_u32() + 1, Ordering::AcqRel, Ordering::Acquire)
```

Comment:
Why do we have to add 1 here? `id` is a `NonZeroU32` right?

Reply:
Oh I see that `as_u32` subtracts 1, that feels a little weird.

Reply:
The `Id` from/to `u32` conversion methods increment/decrement by 1 so we undo that here such that we can keep using `0` as `None`. We could also not do that and use `u32::MAX` as `None`, I have no strong preference here